### PR TITLE
Fix 21785 - Improve error messages for opaque enums

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
The error messages for default initialized variables of opaque enums were horrid and contained misleading informations.
This PR removes them by by checking whether semantic was already run in case of an enum without members.

Some examples:

```d
enum NoBase;
enum WithBase : long;

// Not handled yet
struct S;
enum OpaqueBase : S;

void main()
{
    NoBase nb;
    WithBase wb;
    OpaqueBase ob;
}
```

### CODE REVIEW

**Comment by Geod24** [comment]
Well we don't know the enum definition, so assuming the initializer is its base `.init` is wrong. The enum could start at 1 for example, or `long.min`.

**Comment by MoonlightSentinel** [comment]
> Well we don't know the enum definition, so assuming the initializer is its base `.init` is wrong. The enum could start at 1 for example, or `long.min`.

Aren't opaque enums more intended as a strong typedef (e.g. `__c_long`).

But i guess I'll adapt this PR to create a proper error message.

**Comment by ibuclaw** [comment]
> Aren't opaque enums more intended as a strong typedef (e.g. `__c_long`).
> 
> But i guess I'll adapt this PR to create a proper error message.

Nope, they are just hidden/forward reference types, typically you pass them around by pointer, same as opaque structs.

C long and friends are just a special case.

File: test/fail_compilation/ice10770.d [hunk 1]
```diff
@@ -1,13 +0,0 @@
...
-fail_compilation/ice10770.d(13): Error: enum `ice10770.E2` is forward referenced looking for base type
-fail_compilation/ice10770.d(13):        while evaluating: `static assert(is(E2 e == enum))`
----
-*/
-
-enum E1 : int;
-static assert(is(E1 e == enum) && is(e == int));
-
-enum E2;
-static assert(is(E2 e == enum));
```
Comment (MoonlightSentinel) [inline test/fail_compilation/ice10770.d:13]: This snippet doesn't involve a forward reference and therefore compiles now.

File: src/dmd/denum.d [hunk 2]
```diff
@@ -308,7 +308,10 @@ extern (C++) final class EnumDeclaration : ScopeDsymbol
                 return memtype.defaultInit(loc);
             }
 
-            error(loc, "forward reference of `%s.init`", toChars());
+            if (semanticRun >= PASS.semanticdone)
+                error(loc, "is opaque and has no default initializer");
+            else
+                error(loc, "forward reference of `%s.init`", toChars());
```
Comment (MoonlightSentinel) [inline src/dmd/denum.d:314]: This line isn't covered because actual forward references crash dmd...

Will open another PR for those errors.

**Review by Geod24 — APPROVED** [review]
LGTM. Just a question: What does the error looks like with e.g. `enum NoBase; Nobase nb = void;` ?

**Comment by MoonlightSentinel** [comment]
There is no error because `NoBase` defaults to `int` and `void` initialization is allowed when the base type is not opaque.

**Pull request merged**